File: backend/src/main/kotlin/com/dotcom/retail/domain/catalogue/product/Product.kt
package com.dotcom.retail.domain.catalogue.product

import com.dotcom.retail.common.model.AuditingEntity
import com.dotcom.retail.domain.catalogue.brand.Brand
import com.dotcom.retail.domain.catalogue.category.Category
import com.dotcom.retail.domain.catalogue.image.Image
import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonManagedReference
import jakarta.persistence.*
import org.hibernate.annotations.BatchSize
import org.hibernate.annotations.JdbcTypeCode
import org.hibernate.type.SqlTypes
import java.math.BigDecimal

@Entity
class Product(

    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long = 0,

    var name: String,
    var sku: String,

    var description: String? = null,

    var price: BigDecimal = BigDecimal.ZERO,
    var salePrice: BigDecimal = BigDecimal.ZERO,

    var stock: Int = 0,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "brand_id")
    @JsonManagedReference
    var brand: Brand? = null,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id")
    var category: Category? = null,

    @OneToMany(cascade = [CascadeType.ALL], orphanRemoval = true)
    var images: MutableList<Image> = mutableListOf(),

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(columnDefinition = "jsonb")
    var attributes: MutableMap<String, MutableList<Any>>? = mutableMapOf(),

    var reviewCount: Int = 0,
    var averageRating: Double = 0.0,

    var isActive: Boolean = false,

    @JsonIgnore
    var searchContent: String? = null,

    @Version
    var version: Long? = null

) : AuditingEntity() {

    @PrePersist
    @PreUpdate
    fun updateSearchContent() {
        val attributeText = attributes?.values?.flatten()?.joinToString(" ") { it.toString() } ?: ""

        this.searchContent = """
            $name 
            $sku 
            ${description ?: ""} 
            $attributeText
        """.trimIndent().lowercase().replace("\\s+".toRegex(), " ")
    }

    override fun toString(): String {
        return "Product(id=$id, name='$name', sku='$sku', description=$description, price=$price, salePrice=$salePrice, stock=$stock, attributes=$attributes, isActive=$isActive, ${super.toString()})"
    }
}

File: frontend/src/types/product.ts
import { createPageParams, PageMetadata, PageParams } from "@_types/page"

export interface Product {
  id: number
  name: string
  description: string
  slug: string
  sku: string
  price: number
  salePrice: number
  stock: number
  brand: Brand
  category: Category
  attributes: Attribute[]
  images: Image[]
  averageRating: number
  reviewCount: number
  isActive: true
}

export interface Brand {
  id: number
  name: string
}

export interface Category {
  id: number
  name: string
}

export interface Image {
  id: number
  url: string
  sortOrder: number
  altText: string
}

export type Attribute = {
  name: string
  values: Array<string | number | boolean>
}

export type SortOrder = "TOP" | "PRICE_ASC" | "PRICE_DESC"

export interface ProductQuery extends PageParams {
  categoryId: number
  brands: number[]
  attributes: Attribute[]
  sort: SortOrder
}

export function createProductQuery(): ProductQuery {
  return {
    categoryId: 0,
    brands: [],
    attributes: [],
    sort: "TOP",
    ...createPageParams(),
  }
}

export interface ProductResponse {
  content: Product[]
  page: PageMetadata
}